django/db/models/manager.py

```diff
@@ -242,13 +244,26 @@ class Manager(BaseManager.from_queryset(QuerySet)):
 class ManagerDescriptor(object):
     # This class ensures managers aren't accessible via model instances.
     # For example, Poll.objects works, but poll_obj.objects raises AttributeError.
-    def __init__(self, manager):
+    # It also ensures managers defined on non-abstract base classes are not inherited by child classes.
+    def __init__(self, manager, model):
         self.manager = manager
+        self.model = model
 
     def __get__(self, instance, cls=None):
         if instance is not None:
             raise AttributeError("Manager isn't accessible via %s instances" % cls.__name__)
-        return self.manager
+
+        for model in cls.mro()[:-2]:
+            model_is_concrete = not (model._meta.abstract or model._meta.proxy or model._meta.swapped)
+            if model_is_concrete and cls._meta.concrete_model and model != cls._meta.concrete_model:
+                continue
+            if isinstance(model.__dict__.get(self.manager.name, None), (ManagerDescriptor, AbstractManagerDescriptor)):
+                return self.manager
+
+        msg = ("Access to manager '%s' defined on non-abstract base "
+               "class '%s' from child class '%s' is deprecated."
+               % (self.manager.name, self.model.__name__, cls.__name__))
+        warnings.warn(msg, RemovedInDjango20Warning, stacklevel=2)
```

Comment:
We use hanging indent and try to avoid non-multiple of 4 space indent, something like:
```
warnings.warn(
    "Access to manager '%s' defined on non-abstract base "
    "class '%s' from child class '%s' is deprecated."
    % (self.manager.name, self.model.__name__, cls.__name__),
    ....
)
```
Add something like: "Add the manager to the child class to silence this warning."